btree/btree.c

```diff
@@ -3346,6 +3346,113 @@ static void ff_rec_del_credit(const struct nd *node, m0_bcount_t ksize,
  *	Pros: The performance can be slightly better as it will remove CRC
  *	storage and calculation.
  *	Cons: Any metadata corruption will result in undefined behavior.
+ *
+ *
+ * @brief Proposed design to use benefits of indirect addressing and embedding
+ *        records in nodes of FKVV type.
+ *
+ * In the case of FKVV with the embedded records, we store the key and value on
+ * the left and right locations in the node. But, while performing a write
+ * operation, we need to capture all the records which we moved in the node. To
+ * reduce the movement of records and transaction capturing, we design FKVV node
+ * format which provided embedded records with indirect addressing.
+ *
+ * Fixed key variable value node = {
+ * 	node_header,
+ * 	key = { k1, k2, ...},
+ * 	dir = { orec1, orec2 , orec3, ...},
+ * 	val = { v1, v2,, ...}
+ * }
+ *
+ * We will use following offset directory structure :
+ * struct dir_rec {
+ * 	void *kptr,
+ * 	void *vptr,
+ * 	uint32_t  val_size,
+ * };
+ *
+ *
+ * +------+----+----+----+------+-----+-----+-----+------+-----+----+----+
+ * |      |    |    |    |      |     |     |     |      |     |    |    |
+ * |node  |    |    |    |      |     |     |     |      |     |    |    |
+ * |header| k0 | k1 | k2 +-->   |orec0|orec1|orec2|  <---+ v2  | v1 | v0 |
+ * |      |    |    |    |      |     |     |     |      |     |    |    |
+ * |      |    |    |    |      |     |     |     |      |     |    |    |
+ * +------+----+----+----+------+-----+-----+-----+------+-----+----+----+
+ *        |--key region--|      |----offset dir---|      |--value region-|
+ *
+ * This node format will have keys on left side which will start populating from
+ * left to right whereas values will be present on right side which will be
+ * populated from left to right.
+ * Node will also have offset directory. Initially, this directory will be
+ * situated at the middle of node but if we encounter the possibilty for
+ * overlapping of keys/values with offset directory, we will move the
+ * entire direcory to left or right.
+ *
+ * All directory entries are sorted according to the keys that they point to.
+ * i.e *orec1->kptr < *orec1->kptr < *orec1->kptr
+ *
+ * When inserting a new record, we will find the space between any two records
+ * such that it is able to accomodate new record. Bitmaping can be used to keep
```

Comment:
We do not plan to use bitmap for free records. Instead we will track the total valid records and total max records. All the records between total max records and total valid records contain entries which track the free slots.